cmake/modules/kconfig.cmake

```diff
@@ -146,13 +147,21 @@ else()
   set(_local_TOOLCHAIN_HAS_PICOLIBC n)
 endif()
 
+# Extract git commit information for application
+git_commit_hash(${APPLICATION_SOURCE_DIR} app_git_commit_hash)
+if(app_git_commit_hash)
+	git_commit_hash_short(${app_git_commit_hash} app_git_commit_hash_short app_git_commit_hash_short_int)
+endif()
```

Comment:
initial observation is that this will not work.

Generation of `version.h` is done at build time, and only sources including `version.h` will need to rebuild.
This allows for fast incremental builds in cases where minor source code changes are commited, for example:
- build
- edit `foo.c`
- commit source code changes
- re-build
   - version.h is re-generated
   - foo.c is rebuilt
   - c-files with dependencies to `version.h` are rebuilt. (examples: version.c, banner.c)
   - Zephyr / app is relinked

In today's solution then \ CMake and Kconfig doesn't need to re-run if there is only source code changes.

But introducing git sha into Kconfig to be used as setting or a dependency suddenly means Kconfig must depend on git commits and thus re-run whenever a commit is made. As CMake depends on Kconfig, then that means that CMake also needs to re-run. This creates a new autoconf.h and thus everything must rebuild, which impacts speed of incremental builds.

See also: https://github.com/zephyrproject-rtos/zephyr/issues/40167 and https://github.com/zephyrproject-rtos/zephyr/pull/42527